ggml-vulkan.cpp

```diff
@@ -0,0 +1,5220 @@
+
+        return;
+    }
+#ifdef VK_DEBUG
+    std::cerr << "STAGING" << std::endl;
+#endif
+
+    // Fall back to staging buffer
+    if (vk_staging.size < src->size) {
+        if (!allow_staging_resize) {
+            std::cerr << "ERROR: Staging buffer too small. Required: " << src->size << ", found: " << vk_staging.size << std::endl;
+            GGML_ASSERT(false);
+        }
+        // Resize buffer
+        if (vk_staging.size > 0) {
+            ggml_vk_destroy_buffer(vk_staging);
+        }
+        vk_staging = ggml_vk_create_buffer(src->size, vk::MemoryPropertyFlagBits::eHostVisible | vk::MemoryPropertyFlagBits::eHostCoherent | vk::MemoryPropertyFlagBits::eHostCached);
+    }
+
+    ggml_vk_sync_buffers(ctx.s->buffer, { ggml_vk_subbuffer(*src) }, q, vk::AccessFlagBits::eMemoryRead, vk::AccessFlagBits::eMemoryWrite, true);
+    ctx.s->buffer.copyBuffer(src->buffer, vk_staging.buffer, slices);
+
+    GGML_ASSERT(memcpys != nullptr);
+
+    deferred_memcpy(dst, vk_staging.ptr, dpitch * height, memcpys);
+
+    return;
+}
+
+static void ggml_vk_buffer_read_async(vk_context& ctx, vk_buffer* src, size_t offset, void * dst, size_t size, vk_queue& q, std::vector<vk_staging_memcpy>* memcpys = nullptr) {
+    return ggml_vk_buffer_read_2d_async(ctx, src, offset, dst, size, size, size, 1, q, memcpys);
+}
+
+static void ggml_vk_buffer_read(vk_buffer* src, size_t offset, void * dst, size_t size, vk_queue& q) {
+#ifdef VK_DEBUG
+    std::cerr << "ggml_vk_buffer_read(" << offset << ", " << size << ")" << std::endl;
+#endif
+    if(src->memory_property_flags & vk::MemoryPropertyFlagBits::eHostVisible) {
+        GGML_ASSERT(src->memory_property_flags & vk::MemoryPropertyFlagBits::eHostCoherent);
+
+        memcpy(dst, (uint8_t *) src->ptr + offset, size);
+    } else {
+        vk_context * ctx = ggml_vk_create_context();
+        ggml_vk_ctx_begin(*ctx, q);
+        std::vector<vk_staging_memcpy> staging;
+        ggml_vk_buffer_read_async(*ctx, src, offset, dst, size, q, &staging);
+        ggml_vk_ctx_end(*ctx);
+        ggml_vk_submit(q, ctx->seqs, vk_fence);
+        VK_CHECK(vk_device.device.waitForFences({ vk_fence }, true, uint64_t(-1)), "vk_buffer_read waitForFences");
+        vk_device.device.resetFences({ vk_fence });
+
+        for (auto& cpy : staging) {
+            memcpy(cpy.dst, cpy.src, cpy.n);
+        }
+    }
+}
+
+static void ggml_vk_buffer_copy(vk_buffer * dst, size_t dst_offset, vk_buffer * src, size_t src_offset, size_t size, vk_queue& q) {
+#ifdef VK_DEBUG
+    std::cerr << "ggml_vk_buffer_copy(" << size << ")" << std::endl;
+#endif
+    VkBufferCopy bc{ src_offset, dst_offset, size };
+
+    vk_context * ctx = ggml_vk_create_context();
+    ggml_vk_ctx_begin(*ctx, q);
+    vkCmdCopyBuffer(ctx->s->buffer, src->buffer, dst->buffer, 1, &bc);
+    ggml_vk_ctx_end(*ctx);
+    ggml_vk_submit(q, ctx->seqs, vk_fence);
+    VK_CHECK(vk_device.device.waitForFences({ vk_fence }, true, uint64_t(-1)), "vk_buffer_copy waitForFences");
```

Comment:
Is this supposed to be an infinite wait time? Why are we casting a signed integer to an unsigned integer?, e.g. `uint64_t(-1)`.

Would using `UINT64_MAX` be better? e.g., 

```cpp
VK_CHECK(vk_device.device.waitForFences({ vk_fence }, true, UINT64_MAX), "vk_buffer_copy waitForFences");
```

Reply:
Yeah, I wrote that a long time ago and copied it around. Thank you.